lib/Sema/TypeCheckPattern.cpp

```diff
@@ -123,16 +188,90 @@ lookupUnqualifiedEnumMemberElement(TypeChecker &TC, DeclContext *DC,
 }
 
 /// Find an enum element in an enum type.
-static EnumElementDecl *
-lookupEnumMemberElement(TypeChecker &TC, DeclContext *DC, Type ty,
-                        Identifier name, SourceLoc UseLoc) {
+static EnumElementDecl *lookupEnumMemberElement(TypeChecker &TC,
+                                                DeclContext *DC, Type ty,
+                                                DeclName name,
+                                                SourceLoc UseLoc) {
   if (!ty->mayHaveMembers())
     return nullptr;
 
   // Look up the case inside the enum.
   // FIXME: We should be able to tell if this is a private lookup.
   NameLookupOptions lookupOptions = defaultMemberLookupOptions;
   LookupResult foundElements = TC.lookupMember(DC, ty, name, lookupOptions);
+
+  ArrayRef<Identifier> queryParams = name.getArgumentNames();
+
+  if (foundElements.size() != 1) {
+    // Look up enum element only with base name
+    foundElements =
+        TC.lookupMember(DC, ty, name.getBaseIdentifier(), lookupOptions);
+
+    int numFoundEntries = 0;
+    LookupResultEntry hold = nullptr;
+    LookupResult singleResult;
+    for (LookupResultEntry resultEnt : foundElements) {
+      ValueDecl *e = resultEnt.getValueDecl();
+      if (auto *var = dyn_cast<VarDecl>(e)) {
+        singleResult.add(resultEnt, false);
+        continue;
+      }
+      if (numFoundEntries == 0) {
+        singleResult.add(resultEnt, false);
+        hold = resultEnt;
+      }
+      numFoundEntries++;
+    }
+    if (numFoundEntries == 1)
+      return filterForEnumElement(TC, DC, UseLoc,
+                                  /*unqualifiedLookup=*/false, singleResult);
+
+    LookupResult revisedResult;
+    // Prune results that don't match
+    for (LookupResultEntry resultEnt : foundElements) {
+      ValueDecl *e = resultEnt.getValueDecl();
+      // Maintaining part of list list for filter results
+      if (auto *var = dyn_cast<VarDecl>(e)) {
+        revisedResult.add(resultEnt, false);
+        continue;
+      }
+      DeclName enumElement = resultEnt.getValueDecl()->getFullName();
```

Comment:
You already stored this as `e` above.

```suggestion
      DeclName enumElement = e->getFullName();
```